src/types/context.rs

```diff
@@ -164,14 +164,26 @@ impl Context {
         lock.reassign_non_complete(bound, new);
     }
 
-    /// Binds the type to a given bound. If this fails, attach the provided
-    /// hint to the error.
+    /// Binds the type to a product bound formed by the two inner types. If this
+    /// fails, attach the provided hint to the error.
     ///
     /// Fails if the type has an existing incompatible bound.
-    pub fn bind(&self, existing: &Type, new: Bound, hint: &'static str) -> Result<(), Error> {
+    pub fn bind_product(
+        &self,
+        existing: &Type,
+        prod_l: &Type,
+        prod_r: &Type,
+        hint: &'static str,
+    ) -> Result<(), Error> {
+        assert_eq!(existing.ctx, *self);
+        assert_eq!(prod_l.ctx, *self);
+        assert_eq!(prod_r.ctx, *self);
```

Comment:
d42c496c15925caab8cb55b873b5eb744a9d5598: We should add a readable error message and update the function description with a `# Panics` section.

Reply:
Addressed.